### PR TITLE
Initial AD GenericMaterialProperty problem

### PR DESCRIPTION
**Pull request by vanwdani**
Refs #15213

Follow on PR for new AD template which causes qp index errors.
@bwspenc @lindsayad Do you have any insight?

### CODE REVIEW

**Comment by lindsayad** [comment]
Can you summarize what the issue is that you're seeing?

**Comment by bwspenc** [comment]
@lindsayad When we try to consolidate these classes (which grab components of tensors or do other reductions on them to provide a single Real value from a tensor) into a single templated class that uses `getGenericMaterialProperty()`, it seems like we're grabbing values from the wrong quadrature points. If we dump out the .e file to text (I think using exotxt or ncdump) and diff the new ones with ones generated from the gold files, you can see that we have all the right numbers, they're just reported for different elements.

**Comment by lindsayad** [comment]
Ok @vanwdani @bwspenc 71e978f should hopefully fix things...

**Comment by vanwdani** [comment]
@lindsayad thanks, that worked great.

**Comment by bwspenc** [comment]
@lindsayad Thanks a lot for sorting that out!

@vanwdani Now that this issue is resolved, you should just make those changes we discussed yesterday (to use enumerations for the invariant type) on top of what you have here, rather than fixing it in the version that has two separate classes for AD and non-AD. I guess you could either put it in another commit in this PR or do a separate PR after this is merged, whichever you prefer.

**Comment by vanwdani** [comment]
@bwspenc I pulled this down last night and made the changes we discussed.  I'm pushing it up again, but I think I have a few more revisions I might want to make.

**Comment by lindsayad** [comment]
@bwspenc I am quite confident that those aux variable failures in BISON triggered by the AD 2.0 change was due to this bug.

**Comment by bwspenc** [comment]
@lindsayad I'm having trouble remembering what the AuxVariable problems were in BISON. We must have resolved them somehow. Maybe we should talk about that a little before we merge this.

**Comment by bwspenc** [comment]
So if we have models with and without the `AD` prefix sharing the same implementation like this, do we end up not having a documentation page for the `AD` version? That's what this looks like. That's not completely a bad thing -- we don't want duplicate documentation pages floating around for no reason, but there should be some indication in the documentation system that the `AD` model exists and that the documentation for it is in the non-AD version.

File: modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C:32]: Whoa, hold on here!

Not every one of the invariants makes sense for every possible tensor, which is why we had a limited set defined originally. Some of these can be expanded to allow for other quantities to be computed, but not all of them, and we need to do this in a thoughtful way.  For now, please revert all of them back to the original set. 

Right off the top of my head, these are the obvious ones that could be expanded:
l2norm: It's OK for all of them
max_principal, mid_principal, min_principal: These should also be applicable to the strain
If we're going to add any, we need to test to make sure we've got them right, and add regression tests. For now, let's just leave them alone, and come back later and expand it.

File: modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C:74]: Same deal with all of these. Let's revert them back for now, and potentially add some as we add tests.

File: modules/tensor_mechanics/include/utils/RankTwoScalarTools.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/utils/RankTwoScalarTools.h:48]: This is getting picky, but I think it would be a little more clear if this were renamed to `CYLINDRICAL_COMPONENT`

**Review by bwspenc — CHANGES_REQUESTED** [review]
This looks really good! Just a couple of little requested changes.

Once you address those minor things, I still have two questions from this that I'd like to discuss with the group:
1) Is there anything that these changes (Alex's stuff, specifically) would break in BISON, based on @lindsayad's comment?
2) How do we handle documentation for the AD and non-AD materials?

File: modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C
Comment (vanwdani) [reply modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C:32]: I set all these conditions to quickly convert over input files in BISON.  Files like RIA_CABRI_REP_Na4 have auxvariable/auxkernel pairs that follow the same convention as the already established sets.  (i.e. axial_creep_strain, hoop_elastic_strain, etc. )  I planned on converting over theses pairs, then removing the unnecessary tensors from the MOOSE PR.  I thought it would be faster to remove the tensors, than to collect all the necessary conditions, rebuild MOOSE, rebuild BISON, and then change the input files.

**Comment by vanwdani** [comment]
@bwspenc I followed the no-AD convention of ComputeIsotropicElasticityTensor.md. This is a class that uses <is_ad> and the documention didn't mention AD.  I'm not sure if that was right or wrong

**Comment by lindsayad** [comment]
> @lindsayad I'm having trouble remembering what the AuxVariable problems were in BISON. We must have resolved them somehow. Maybe we should talk about that a little before we merge this.

We saw exodiffs in a few tests so so we loosened the tolerances

**Comment by bwspenc** [comment]
@lindsayad Well, if slightly loosening tolerances got them passing, then this shouldn't adversely affect them -- they'll just match the original solutions better. This is clearly an important fix.

File: modules/tensor_mechanics/doc/content/source/materials/ADRankTwoInvariant.md
Comment (bwspenc) [inline modules/tensor_mechanics/doc/content/source/materials/ADRankTwoInvariant.md:7]: Umm, I'm pretty sure you use 'an' before a word that starts with an vowel.

I'll not hold up this PR for this, though.  Let's keep the AD and non-AD pages here for the time being, assuming the documentation builds fine, but we need to figure out a better long-term solution.

**Review by bwspenc — APPROVED** [review]

**Comment by bwspenc** [comment]
The AD doc pages do show up and appear to be linked with AD classes, so until we come up with something better, I guess the thing to do is have doc pages for both classes.

File: modules/tensor_mechanics/doc/content/source/materials/ADRankTwoInvariant.md
Comment (vanwdani) [reply modules/tensor_mechanics/doc/content/source/materials/ADRankTwoInvariant.md:7]: Ha!  You are right.  I just added AD into the duplicate non-AD version.  It would make more sense to have the one md file that describes it's ability to do both things.

File: modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C
Comment (bwspenc) [reply modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C]: I think those ones you mentioned make sense. Go ahead and submit another PR that adds the specific cases that are being used. I just don't think we can quite go and wholesale add every combination of options.

**Pull request merged**
